OpenDreamClient/Rendering/DreamViewOverlay.cs

```diff
@@ -103,84 +435,471 @@ sealed class DreamViewOverlay : Overlay {
+    private void ReturnRenderTarget(IRenderTexture rental) {
         if (!_renderTargetCache.TryGetValue(rental.Size, out var storeList))
             storeList = new List<IRenderTexture>(4);
 
         storeList.Add(rental);
         _renderTargetCache[rental.Size] = storeList;
     }
 
-    // TODO: Move this to DreamIcon.Draw() so screen objects can have filters
-    private void DrawIcon(DrawingHandleWorld handle, DreamIcon icon, Vector2 position) {
-        if (icon.Appearance == null)
-            return;
+    private void ClearRenderTarget(IRenderTexture target, DrawingHandleWorld handle, Color clearColor)
+    {
+        handle.RenderInRenderTarget(target, () => {}, clearColor);
+    }
 
-        position += icon.Appearance.PixelOffset / (float)EyeManager.PixelsPerMeter;
 
-        foreach (DreamIcon underlay in icon.Underlays) {
-            DrawIcon(handle, underlay, position);
+    private ShaderInstance GetBlendAndColorShader(RendererMetaData iconMetaData, Color? colorOverride = null, int? blendOverride = null)
+    {
+        Color RGBA = colorOverride == null ? iconMetaData.ColorToApply.WithAlpha(iconMetaData.AlphaToApply) : colorOverride.Value;
+        ColorMatrix colorMatrix;
+        if(colorOverride != null || iconMetaData.ColorMatrixToApply == null)
+            colorMatrix = new ColorMatrix(RGBA);
+        else
+            colorMatrix = iconMetaData.ColorMatrixToApply.Value;
+        ShaderInstance blendAndColor;
+        if(blendOverride != null || !_blendmodeInstances.TryGetValue(iconMetaData.BlendMode, out blendAndColor))
+            blendAndColor = _blendmodeInstances[blendOverride == null ? 0 : blendOverride.Value];
+        blendAndColor = blendAndColor.Duplicate();
+        blendAndColor.SetParameter("colorMatrix", colorMatrix.GetMatrix4());
+        blendAndColor.SetParameter("offsetVector", colorMatrix.GetOffsetVector());
+        return blendAndColor;
+    }
+
+    private (Action, Action) DrawIconAction(DrawingHandleWorld handle, RendererMetaData iconMetaData, Vector2 positionOffset, Texture textureOverride = null) {
+        DreamIcon icon = iconMetaData.MainIcon;
+
+        Vector2 position = iconMetaData.Position + positionOffset;
+        Vector2 pixelPosition = position*EyeManager.PixelsPerMeter;
+
+        Texture frame;
+        if(textureOverride != null) {
+            frame = textureOverride;
+            //we flip this because GL's coordinate system is bottom-left first, and so render target textures are upside down
+            iconMetaData.TransformToApply = iconMetaData.TransformToApply * flipMatrix;
+        }
+        else
+            frame = icon.CurrentFrame;
+
+        //KEEP_TOGETHER groups
+        if(iconMetaData.KeepTogetherGroup != null && iconMetaData.KeepTogetherGroup.Count > 0)
+        {
+            //store the parent's transform, color, blend, and alpha - then clear them for drawing to the render target
+            Matrix3 KTParentTransform = iconMetaData.TransformToApply;
```

Comment:
All these variables don't match the naming convention for locals (`KTParentTransform` should be `ktParentTransform`)

- KTParentTransform
- KTParentColor
- KTParentAlpha
- KTParentBlendMode
- KTItems
- TempTexture
- KTItem
- KTTexture